### PR TITLE
Add benchmark for hash vs btree dictionary

### PR DESCRIPTION
**Pull request by baldersheim**
@toregge or @geirst PR

### CODE REVIEW

**Review by toregge — APPROVED** [review]

File: tests/performance/lookup/lookup.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      }
+    ]
+    num_docs = 1000000
+    num_queries = num_docs
+    upper_limit = num_docs*10
+    keys_per_query = 100
+    num_clients = 16
+    deploy_app(get_app())
+    container = (vespa.qrserver["0"] or vespa.container.values.first)
+    container.execute("g++ -Wl,-rpath,#{Environment.instance.vespa_home}/lib64/ -g -O3 -o #{dirs.tmpdir}/docs #{selfdir}/docs.cpp")
```
Comment (toregge) [inline tests/performance/lookup/lookup.rb:56]: This assumes that the program source exists on the container in the same directory as on the performance test controller node.

File: tests/performance/lookup/query.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    for (unsigned int i(1); i < keys_per_query; i++) {
+        keys << ",%22" << (rand()%upper_limit) << "%22%3A1";
+    }
+    const char * queryPrefix = "/search/?yql=select%20*%20from%20sources%20*%20where%20weightedSet";
+    printf("%s(%s,%7B%s%7D)%3B\n", queryPrefix, field, keys.str().c_str());
+}
+
+int
+main(int argc, char **argv) {
+    long numQueries = atoi(argv[1]);
```
Comment (geirst) [inline tests/performance/lookup/query.cpp:19]: Consider calling srand() to produce same random data between runs.

File: tests/performance/lookup/lookup.rb (hunk 1 above)
Comment (geirst) [inline tests/performance/lookup/lookup.rb:56]: As a followup, we should create a helper function to compile C++ programs. This is spread across many performance tests now.

**Review by geirst — APPROVED** [review]
👍

File: tests/performance/lookup/docs.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#include <cstdlib>
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply tests/performance/lookup/docs.cpp:1]: Fixed

File: tests/performance/lookup/lookup.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
+# Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply tests/performance/lookup/lookup.rb:1]: Fixed

File: tests/performance/lookup/query.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#include <cstdlib>
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply tests/performance/lookup/query.cpp:1]: Fixed

File: tests/performance/lookup/test.sd [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply tests/performance/lookup/test.sd:1]: Fixed

File: tests/performance/lookup/lookup.rb [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+require 'performance_test'
+require 'app_generator/search_app'
+require 'performance/fbench'
+require 'environment'
+
+class Lookup < PerformanceTest
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply tests/performance/lookup/lookup.rb:7]: Fixed

File: tests/performance/lookup/lookup.rb (hunk 1 above)
Comment (baldersheim) [reply tests/performance/lookup/lookup.rb:56]: I assume it does as I follow the pattern seen in other tests.
But I agree, not the perfect solution.

Comment (baldersheim) [reply tests/performance/lookup/lookup.rb:56]: Agree, I will keep that in mind.

File: tests/performance/lookup/query.cpp (hunk 2 above)
Comment (baldersheim) [reply tests/performance/lookup/query.cpp:21]: Done. Note that the seed is 1 if srand is never called.

**Pull request merged**